File: src/main/java/at/yawk/katbot/RssFeedListener.java
package at.yawk.katbot;

import com.rometools.rome.feed.synd.SyndEntry;
import com.rometools.rome.feed.synd.SyndFeed;
import com.rometools.rome.io.FeedException;
import com.rometools.rome.io.SyndFeedInput;
import java.io.IOException;
import java.io.InputStream;
import java.net.URI;
import java.time.Instant;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import javax.inject.Inject;
import javax.inject.Provider;
import javax.inject.Singleton;
import lombok.Value;
import lombok.extern.slf4j.Slf4j;
import org.apache.http.HttpResponse;
import org.apache.http.client.HttpClient;
import org.apache.http.client.methods.HttpGet;
import org.kitteh.irc.client.library.Client;
import org.xml.sax.InputSource;

/**
 * @author yawkat
 */
@Singleton
@Slf4j
public class RssFeedListener {
    @Inject Client client;
    @Inject ScheduledExecutorService executor;
    @Inject Config config;
    @Inject HttpClient httpClient;
    // lazy init
    @Inject Provider<UrlShortener> urlShortener;

    private final SyndFeedInput input = new SyndFeedInput();

    private final Map<URI, Instant> lastPollTimes = new HashMap<>();

    private SyndFeed loadFeed(URI uri) throws IOException, FeedException {
        HttpResponse response = httpClient.execute(new HttpGet(uri));
        try (InputStream in = response.getEntity().getContent()) {
            return input.build(new InputSource(in));
        }
    }

    private synchronized void poll() throws Exception {
        for (Map.Entry<URI, FeedConfiguration> conf : config.getFeeds().entrySet()) {
            URI uri = conf.getKey();

            SyndFeed feed = loadFeed(uri);
            Instant deadline = lastPollTimes.get(uri);
            Instant newDeadline = null;
            for (SyndEntry feedEntry : feed.getEntries()) {
                Instant entryTime = feedEntry.getPublishedDate().toInstant();
                if (newDeadline == null || entryTime.isAfter(newDeadline)) {
                    newDeadline = entryTime;
                }
                if (deadline != null && entryTime.isAfter(deadline)) {
                    fire(conf.getValue(), feedEntry);
                }
            }
            lastPollTimes.put(uri, newDeadline);
        }
    }

    private void fire(FeedConfiguration conf, SyndEntry feedEntry) {
        String message = Template.parse(conf.messagePattern)
                .set("title", feedEntry.getTitle())
                .set("uri", feedEntry.getUri())
                .set("uri.short", () -> {
                    try {
                        return urlShortener.get().shorten(URI.create(feedEntry.getUri())).toString();
                    } catch (Exception e) {
                        throw new RuntimeException(e);
                    }
                })
                .finish();

        log.info("Sending feed update '{}' to {} channels", message, conf.channels.size());

        ForumListener.sendToChannels(client, conf.channels, message);
    }

    public void start() {
        executor.scheduleWithFixedDelay(() -> {
            try {
                poll();
            } catch (Exception e) {
                log.error("Failed to poll RSS feeds", e);
            }
        }, 0, 1, TimeUnit.MINUTES);
    }

    @Value
    public static class FeedConfiguration {
        private final List<String> channels;
        private final String messagePattern;
    }
}

File: src/main/kotlin/at/yawk/katbot/RssFeedListener.kt
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package at.yawk.katbot

import com.rometools.rome.feed.synd.SyndEntry
import com.rometools.rome.feed.synd.SyndFeed
import com.rometools.rome.io.SyndFeedInput
import org.apache.http.client.HttpClient
import org.apache.http.client.methods.HttpGet
import org.slf4j.LoggerFactory
import org.xml.sax.InputSource
import java.net.URI
import java.time.Instant
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Provider

/**
 * @author yawkat
 */
class RssFeedListener @Inject constructor(
        val ircProvider: IrcProvider,
        val executor: ScheduledExecutorService,
        val config: Config,
        val httpClient: HttpClient,
        // lazy init
        val urlShortener: Provider<UrlShortener>
) {
    companion object {
        private val log = LoggerFactory.getLogger(RssFeedListener::class.java)
    }

    val input = SyndFeedInput()
    val lastPollTimes = hashMapOf<URI, Instant>()

    private fun loadFeed(uri: URI): SyndFeed {
        val response = httpClient.execute(HttpGet(uri))
        response.entity.content.use { `in` -> return input.build(InputSource(`in`)) }
    }

    @Synchronized
    private fun poll() {
        for (conf in config.feeds.entries) {
            val uri = conf.key

            val feed = loadFeed(uri)
            val deadline = lastPollTimes[uri]
            var newDeadline: Instant? = null
            for (feedEntry in feed.entries) {
                val entryTime = feedEntry.publishedDate.toInstant()
                if (newDeadline == null || entryTime.isAfter(newDeadline)) {
                    newDeadline = entryTime
                }
                if (deadline != null && entryTime.isAfter(deadline)) {
                    fire(conf.value, feedEntry)
                }
            }
            if (newDeadline != null) {
                lastPollTimes[uri] = newDeadline
            }
        }
    }

    private fun fire(conf: FeedConfiguration, feedEntry: SyndEntry) {
        val message = conf.messagePattern
                .with("title", feedEntry.title)
                .with("uri", feedEntry.uri)
                .with("uri.short", urlShortener.get().shorten(URI.create(feedEntry.uri)).toString())

        log.info("Sending feed update '{}' to {} channels", message, conf.channels.size)

        message.sendTo(ircProvider.findChannels(conf.channels))
    }

    fun start() {
        executor.scheduleWithFixedDelay({
            try {
                poll()
            } catch (e: Exception) {
                log.error("Failed to poll RSS feeds", e)
            }
        }, 0, 1, TimeUnit.MINUTES)
    }

    data class FeedConfiguration(
            val channels: List<String>,
            val messagePattern: at.yawk.katbot.template.Template
    )
}